### PR TITLE
feat(NES-1644): template gallery preview revalidate + custom-domain gate

### PR DESCRIPTION
**Pull request by csiyang**
## Summary

NES-1644 absorbed NES-1643. The shipped scope is narrower than the original plan — see the plan doc's Outcome section for the full evolution. Final state:

- **Public template-gallery page: SSR + uncached resolver.** `apps/journeys/pages/home/template-gallery/[slug].tsx` uses `getServerSideProps` with `Cache-Control: no-store, max-age=0`. `Query.templateGalleryPageBySlug` is set to TTL 0 in `apis/api-journeys-modern/src/yoga.ts`. Replaces the original ISR + revalidate-chain design — see the runtime-errors solution doc below for why.
- **Authenticated preview proxy.** `apps/journeys-admin/pages/api/preview-template-gallery.ts` gates Preview behind Firebase auth + `Sec-Fetch-Site` CSRF check + slug validation, then 307-redirects to the public URL. The only proxy that survived the simplification.
- **Custom-domain publish gate.** `useCanPublishCollection` hook returns `{ canPublish, reason, loading }`. Blocks Publish + Preview when the active team has any `customDomain.routeAllTeamJourneys === true`. Tooltip copy verbatim per spec. Loading: fail-open (UX). Error: fail-CLOSED (safety) with distinct copy. Exported for NES-1637 reuse. Console-warn is throttled by team+error message.
- **Trashing or archiving a journey severs its collection membership.** After `journeysTrash` / `journeysArchive` resolves, the dialog fires a best-effort `templateGalleryPageAssignJourney({ journeyId, pageId: null })` to remove the `TemplateGalleryPageTemplate` join row. Without this, restoring a trashed/archived journey would put it back in its prior collection slot (resolver hid it but the join row survived).
- **Apollo cache hygiene across the `Journey` / `TemplateGalleryItem` typename boundary.** The `templates` field on `TemplateGalleryPage` stores entities as `TemplateGalleryItem:<id>` (a Pothos variant), while `journeysTrash` / `journeysArchive` return `Journey:<id>` — Apollo's normal entity-merge can't bridge that. Both dialogs now run an explicit `cache.modify` pass that enumerates every cached `TemplateGalleryPage` and filters the trashed/archived ref out of its `templates` list, then evicts the relevant entity flavour(s). Replaces the prior `cache.evict + cache.gc()` shape which worked locally but was unreliable on stage.
- **Status-filter on the active view's template list.** `TemplateGalleryPageList` re-applies the same `STATUS_FILTER_TO_JOURNEY_STATUSES[status]` predicate the server uses, on the client. Defends against Apollo's normalized cache merging a post-mutation status flip into a ref that the cached list still holds — without this, a just-archived journey would briefly render in the "All Templates" section of the active view.
- **`dragInFlight` is now a ref** (synchronous gating against re-entrant drops) with a mirror state for rendering. `handlePublish` got a `mountedRef` guard. Stale ISR / revalidate references scrubbed from comments + docs.

## Plan + solution docs

- `docs/plans/2026-05-07-001-feat-nes-1644-template-gallery-preview-revalidate-custom-domain-plan.md` — original plan
- `docs/plans/2026-05-14-001-fix-template-gallery-revalidate-cache-invalidation-plan.md` — investigation that pivoted the design (with Outcome section)
- `docs/solutions/runtime-errors/yoga-response-cache-null-stickiness-and-zombie-process-debugging-nes1644.md` — full debugging history (cached-null stickiness in `@graphql-yoga/plugin-response-cache`, the zombie-process trap on port 4004, the evict-vs-modify evolution on the admin side)
- `docs/solutions/best-practices/template-gallery-page-collections-patterns-nes1539.md` — Patterns 14 (SSR + TTL 0), 15 (preview proxy auth posture), 16 (custom-domain gate contracts), 17 (trash/archive sever collection membership), 18 (client-side status predicate)

## Test plan

- [x] `apps/journeys-admin/pages/api/preview-template-gallery.spec.ts` — auth, CSRF, slug validation, 307 redirect
- [x] `apps/journeys-admin/src/libs/useCanPublishCollection/useCanPublishCollection.spec.tsx` — fail-open loading, fail-closed error, routeAll detection, log throttling
- [x] `apps/journeys-admin/src/components/JourneyList/JourneyCard/JourneyCardMenu/TrashJourneyDialog/TrashJourneyDialog.spec.tsx` — cache filtering across multiple pages, dual-entity eviction, unassign pairing, best-effort failure mode
- [x] `apps/journeys-admin/src/components/JourneyList/JourneyCard/JourneyCardMenu/DefaultMenu/ArchiveJourney/ArchiveJourney.spec.tsx` — parallel coverage for archive
- [x] `apps/journeys-admin/src/components/TemplateGalleryPageList/TemplateGalleryPageList.spec.tsx` — regression: archived journey excluded from active view
- [x] `apps/journeys-admin/src/components/TemplateGalleryPageList/CollectionDialog/CollectionPreviewPane/CollectionPreviewPane.spec.tsx` — preview disabled states (no slug, no canPublish)
- [x] `apps/journeys-admin/src/components/TemplateGalleryPageList/CollectionPublishSuccessDialog/CollectionPublishSuccessDialog.spec.tsx` — proxy URL, gate, disabled states
- [x] `apps/journeys/__tests__/pages/home/template-gallery/[slug].spec.tsx` — Cache-Control header, null-branch diagnostic log + 5-error truncation cap
- [x] `apps/journeys/proxy.spec.ts` — `/template-gallery/*` short-circuit across hosts, query preservation, no false-prefix match
- [x] `nx lint journeys` — 0 errors
- [x] `nx lint journeys-admin` — 0 errors
- [ ] Manual: archive a journey that's in a collection on stage. Confirm the journey disappears from "All Templates" immediately (no refresh), the collection card no longer shows it, and unarchiving returns it to "All Templates" rather than its prior collection slot.
- [ ] Manual: same flow for trash.
- [ ] Manual: confirm a custom-domain team with `routeAllTeamJourneys === true` sees Publish + Preview disabled with the tooltip on every Collection surface.
- [ ] Manual: publish a collection from a non-custom-domain team, click Preview, observe the just-published content with no stale-cache window.

## Follow-ups

- **Migrate `journeysTrash` and `journeysArchive` to api-journeys-modern** (Pothos), delete the `TemplateGalleryPageTemplate` join rows transactionally inside each mutation. Frontend `…assignJourney({ pageId: null })` becomes redundant and can be removed. Follows Mike's `chore: *modern` PR pattern. Separate ticket.
- **NES-1637 Share Again** — reuses `useCanPublishCollection` from this PR.

## Deferred (separate tickets)

- **NES-1668** drag-between-drafts cache bug (pre-existing from NES-1539)
- **NES-1669** "journey trashed" snackbar copy on template delete
- **NES-1670** UX feedback for the post-mutation reconciliation window

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/journeys-admin/pages/api/preview-template-gallery.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    if (tokens == null) {
+      return res.status(403).json({ error: 'Not authorized' })
+    }
+  } catch (e) {
+    return res.status(403).json({ error: 'Not authorized' })
+  }
+
+  const slug = req.query.slug?.toString()
+
+  if (slug == null) return res.status(400).json({ error: 'Missing Slug' })
```
[the remark this answers is not in the document]
Comment (csiyang) [reply apps/journeys-admin/pages/api/preview-template-gallery.ts]: Already addressed earlier in this PR — current code does the `typeof` gate plus the stronger `SLUG_RE = /^[a-z0-9]+(-[a-z0-9]+)*$/` regex check at `apps/journeys-admin/pages/api/preview-template-gallery.ts:36-40`. Anti-path-traversal + array-input both rejected.

File: apps/journeys-admin/src/components/TemplateGalleryPageList/CollectionCard/CollectionCard.spec.tsx [hunk 2]
```diff
@@ -196,6 +196,93 @@ describe('CollectionCard', () => {
...
+    )
+    await userEvent.click(
+      screen.getByRole('button', { name: 'Collection actions' })
+    )
+    // Published collections show Unpublish + Preview (no Publish item).
+    expect(screen.getByRole('menuitem', { name: 'Preview' })).toHaveAttribute(
+      'aria-disabled',
+      'true'
+    )
+  })
```
[the remark this answers is not in the document]
Comment (csiyang) [reply apps/journeys-admin/src/components/TemplateGalleryPageList/CollectionCard/CollectionCard.spec.tsx:268]: Good catch — both tests now assert the tooltip text matches the gate reason, not just `aria-disabled`. Hovering the `parentElement` (MUI's `<span>` wrapper around the disabled `<li>`) since the LI has `pointer-events: none`. Fixed in e734c94d5.

File: apps/journeys-admin/src/components/TemplateGalleryPageList/CollectionCard/CollectionCard.tsx [hunk 3]
```diff
@@ -36,13 +51,35 @@ function CollectionCardImpl({
...
   const [ungroupOpen, setUngroupOpen] = useState(false)
 
   const isPublished = collection.status === TemplateGalleryPageStatus.published
+  const isEmpty = collection.templates.length === 0
+
+  // Preview (a) only makes sense once the page is publicly reachable, and
+  // (b) is also gated by custom-domain teams — same constraint as Publish.
+  const previewDisabled = !isPublished || !canPublish
+  const previewHref = `/api/preview-template-gallery?slug=${encodeURIComponent(collection.slug)}`
+  const previewTooltip = !canPublish
```
[the remark this answers is not in the document]
Comment (csiyang) [reply apps/journeys-admin/src/components/TemplateGalleryPageList/CollectionCard/CollectionCard.tsx]: Added `!collection.slug` to `previewDisabled` for UX parity with the server-side `SLUG_RE` check — Preview now disables in the menu rather than opening a tab to a 400. Fixed in e734c94d5.

File: apps/journeys/pages/api/revalidate-template-gallery.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  try {
+    // Public template-gallery page lives at /home/template-gallery/<slug>
+    // (Next rewrite maps your.nextstep.is/template-gallery/<slug> → here).
+    // Custom domains can't host gallery pages (NES-1644), so no hostname branch.
+    const path = `/home/template-gallery/${req.query.slug as string}`
+    await res.revalidate(path)
+
+    return res.status(200).json({
+      revalidated: true
+    })
```
[the remark this answers is not in the document]
Comment (csiyang) [reply apps/journeys/pages/api/revalidate-template-gallery.ts]: Same as the preview endpoint — already gated by `typeof slug !== 'string'` + `SLUG_RE` check at `apps/journeys/pages/api/revalidate-template-gallery.ts:21-23`. Empty + array + path-traversal all rejected.

**Review by mikeallisonJS — CHANGES_REQUESTED** [review]
## Summary

Two well-scoped features: ISR cache revalidation proxy routes for template gallery preview, and a `useCanPublishCollection` hook gating Publish/Preview behind a custom-domain check. Also fixes a real pre-existing React StrictMode bug with `mountedRef`. Test coverage is thorough and security-aware throughout.

---

## 🔴 Critical — Request Changes

### 1. `useCanPublishCollection.ts` line 54–55 — `canPublish: false` while `teamId` is null breaks fail-open contract

```ts
if (teamId == null) {
  return { canPublish: false, reason: null, loading: false }
}
```

The JSDoc on line 41–44 says the gate is **fail-open** ("treats the team as `canPublish: true` to keep the publish CTA enabled by default"). But the `teamId == null` early-return contradicts this: since `teamId` arrives asynchronously from `useTeam()`, there is a transient window on every page load where every team's Publish/Preview button is **disabled** — including legitimate non-custom-domain teams.

The interface comment (line 31–33) says "no team, no publish" but that describes a permanent no-team state, not the loading state. These two cases need to be distinguished.

**Fix:** Return `{ canPublish: true, reason: null, loading: true }` when `teamId` is null so the button stays enabled while the team context resolves. Gate on `teamId == null` only after you've confirmed the team context has settled (e.g. expose a `teamLoading` prop or check the parent's loading state before calling this hook).

---

## 🟡 Medium — Should Fix

### 2. `apps/journeys/pages/api/revalidate-template-gallery.ts` — Timing side-channel on access token comparison

```ts
req.query.accessToken !== process.env.JOURNEYS_REVALIDATE_ACCESS_TOKEN
```

Plain string equality is not constant-time. Use `crypto.timingSafeEqual` for shared-secret comparisons:

```ts
import { timingSafeEqual } from 'crypto'

const provided = Buffer.from(String(req.query.accessToken))
const expected = Buffer.from(process.env.JOURNEYS_REVALIDATE_ACCESS_TOKEN ?? '')
if (provided.length !== expected.length || !timingSafeEqual(provided, expected)) {
  return res.status(401).json({ error: 'Invalid token' })
}
```

### 3. `apps/journeys-admin/pages/api/revalidate-template-gallery.ts` — Upstream HTTP status forwarded verbatim to client

```ts
return res.status(response.status).json({ error: 'Error revalidating' })
```

If the upstream returns `401` (e.g. wrong token in prod), the browser receives a `401`, leaking internal topology. Normalise non-2xx upstream responses to a fixed `502`:

```ts
return res.status(502).json({ error: 'Error revalidating' })
```

The body is already sanitised correctly — only the status code needs fixing.

---

## 💬 Suggestions

### 4. `useRevalidateTemplateGallery.ts` — Not actually a hook

This file contains no `use*` calls and returns a plain async function. The `use` prefix implies React rules-of-hooks apply (can't call conditionally, etc.) for no benefit. Either rename to `revalidateTemplateGallery` as a plain export, or wrap the returned function in `useCallback` so the name earns its prefix and the identity is stable across renders.

### 5. `SLUG_RE` regex duplicated across three files

`/^[a-z0-9]+(-[a-z0-9]+)*$/` is copy-pasted into both admin routes and the journeys public route. Extract into a shared lib constant — copy-pasted security-sensitive regexes create drift risk.

### 6. `CollectionPreviewPane.tsx` — Tooltip fires on enabled button

`viewTooltip` renders "Open in new tab" even when the button is enabled, producing a redundant tooltip that duplicates the `aria-label`. Pass `title=""` when `!previewDisabled`, consistent with how `CollectionCard` handles it.

---

## ✅ What Looks Good

- **Defense-in-depth on the admin proxy is excellent** — POST-only → CSRF header → Firebase auth → slug regex → sanitised error body (the comment explaining *why* the upstream body is suppressed to avoid leaking the access-token query param is commendable).
- - **`SLUG_RE` is correct and well-tested** — rejects path traversal, query injection, uppercase, spaces, and empty strings.
- - - **StrictMode `mountedRef` fix is correct and important** — cleanup-only `useEffect` left ref at `false` after double-invoke in dev; adding setup body assignment is the right fix.
- - - - **Revalidation scoped to published collections only** — `unpublish` and `ungroup` both check `status === published` before firing.
- - - - - **Race-condition guard in `useCollectionForm`** — checking both pre- and post-mutation status with OR is solid.
- - - - - - **`CUSTOM_DOMAIN_PUBLISH_BLOCKED_COPY` exported as a constant** — tooltip copy stays in sync across all surfaces.
- - - - - - - **Test quality is well above baseline** — each test defends a real edge case, not boilerplate.
---

## Verdict

**Request Changes** — issue #1 (`canPublish: false` while `teamId` is null) is a one-line fix that needs to land before merge; it directly contradicts the stated fail-open contract and will visibly flash disabled Publish buttons for all users on load. Issues #2 and #3 are also worth addressing given the security sensitivity of the access token flow. Everything else is clean.

**Comment by csiyang** [comment]
## Code-review fixes applied — commit `4ef5cc938`

This commit applies the P0/P1/P2 findings from a `/ce:review` pass on tip `37d5b4b27`. Summary:

**P0**
- `useCollectionForm.ts` — skip the slug update when the user cleared the field (would otherwise rename the published page's slug to empty and break external links)
- `useCollectionMutations.ts` — `showError` on `{ data: null }` mutation resolutions (publish / unpublish / ungroup)

**P1**
- `revalidate-template-gallery` (both proxies + handler) — token moved from URL query string to `Authorization: Bearer` header; legacy `?accessToken=` form rejected outright so there's exactly one live auth path during migration
- `preview-template-gallery.ts` — restrict to GET only + `Sec-Fetch-Site === same-origin | none` gate (mirrors revalidate proxy posture; closes CSRF gap)
- `useCollectionMutations.ts` — `submittingRef` synchronous guard (closes stale-closure window on two-row simultaneous clicks)
- `useCanPublishCollection.ts` — fail-CLOSED on query error with a distinct "couldn't check" reason + `console.warn` so operators see silent prod failures
- `useCustomDomainsQuery.tsx` — rename `hostname` → `primaryHostname` with JSDoc warning about routeAll semantics; 7 consumers + 2 specs updated via destructure alias

**P2-B**
- `[slug].tsx` `getStaticProps` null-branch log — only fire on errors (not on legit not-found); cap at 5 errors with `truncated` flag; whitelist `extensions.code` only (never log raw `extensions` — can leak server stack traces / Prisma context)

**Doc**
- Extended `docs/solutions/best-practices/template-gallery-page-collections-patterns-nes1539.md` with Patterns 14 (revalidate-on-mutation chain), 15 (`timingSafeEqual` + CSRF), 16 (custom-domain publish gate)

**Deferred**
- P1-D (TrashJourneyDialog optimistic-onTrashSuccess) — separate follow-up ticket
- P2-A (server-side custom-domain reject) — NES-1677 scope
- P2-C/E/F/G/H/I — non-blocker; not in this round

**Local verify**
- 212 specs pass across touched surfaces (190 jest + 22 vitest)
- type-check passes for both apps
- journeys lint passes; journeys-admin lint OOMs locally on full project run — CI runs with adequate heap

NES-1677 (PR #9217, backend cache-invalidation) is the complementary backend half.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by mikeallisonJS — CHANGES_REQUESTED** [review]
## Code Review: `feat(NES-1644): template gallery preview revalidate + custom-domain gate`

Well-engineered PR overall. The ISR → SSR + TTL 0 pivot is the right architectural call, the proxy CSRF design is correct, and the plan/solution docs are excellent institutional knowledge. Two items need attention before merge.

---

### 🔴 Critical

**Success snackbar is blocked by the `unassignFromCollection` await — `ArchiveJourney.tsx` and `TrashJourneyDialog.tsx`**

In both dialogs, `onCompleted` `await`s the unassign call _before_ showing the success snackbar. The PR description calls this "best-effort," but the current code blocks the UX on a secondary network round-trip (~1–2s). The snackbar should fire immediately; the unassign should be fire-and-forget:

```ts
// Current — blocks snackbar on unassign latency
onCompleted: async () => {
  await unassignFromCollection({ variables: { journeyId: id, pageId: null } })
    .catch((e) => console.warn('[ArchiveJourney] failed to unassign', e))
  enqueueSnackbar(t('Journey Archived'), { variant: 'success' })
}

// Fix — snackbar fires immediately, unassign is non-blocking
onCompleted: async () => {
  enqueueSnackbar(t('Journey Archived'), { variant: 'success' })
  unassignFromCollection({ variables: { journeyId: id, pageId: null } })
    .catch((e) => console.warn('[ArchiveJourney] failed to unassign', e))
}
```

The existing test (`waitFor(() => expect(unassignResult).toHaveBeenCalled())`) doesn't catch this because `waitFor` retries until true — it doesn't assert ordering. A test verifying the snackbar appears _before_ the unassign completes would close this gap.

---

### 🟡 Warnings

**`templateGallerySlug.spec.ts` is flagged as binary in the diff**

The diff shows: `Binary files /dev/null and b/libs/journeys/ui/src/libs/templateGallerySlug/templateGallerySlug.spec.ts differ`. A `.spec.ts` file should never be binary — this likely means CRLF line endings were committed on Windows, causing Git to treat it as binary. If so, tests in that file may be silently skipped in CI. Verify with `pnpm nx test journeys-ui` and run `git add --renormalize` + recommit if needed.

**Slug validation should move before Firebase auth in the preview proxy**

Order in `preview-template-gallery.ts`: CSRF check → env check → Firebase token verify → slug validation. Moving the cheap slug check before Firebase avoids an unnecessary token round-trip for obviously malformed slugs. Not a security issue, but a minor latency and defence-in-depth win.

**Shared `submittingRef` silently drops concurrent mutations with no user feedback**

`submittingRef` guards all three mutation functions. A slow publish silently drops a concurrent unpublish on a different collection — no snackbar or spinner change signals this. A brief toast ("Please wait for the current action to finish") when the guard fires would make the drop visible to users.

**`dragInFlight` state and `dragInFlightRef` are managed independently**

Both must always be updated together (`setDragInFlight(true)` + `dragInFlightRef.current = true`). If a future code path sets one without the other, ref and rendered state will silently diverge. A small encapsulating setter or custom hook would enforce the invariant.

**Duplication in the `cache.modify` logic between `TrashJourneyDialog` and `ArchiveJourney`**

The ~40-line `update(cache, { data })` callback is nearly identical in both files. Extracting `evictFromTemplateGalleryPages(cache, journeyIds, { evictJourneyEntity })` would reduce future drift risk.

---

### 🟢 Suggestions

- `handleView` in `CollectionPreviewPane` redundantly checks `slug == null` when `viewDisabled` already encodes that guard.
- `publishBlockedReason` from `useCanPublishCollection` requires every caller to wrap it in `t()`. Consider noting this in the return type's JSDoc.
- The `setTimeout(resolve, 200)` pattern in the `useCollectionMutations` test is brittle under CI load — `jest.useFakeTimers()` would make it deterministic.

---

### ✅ What's Done Well

- **ISR → SSR is the right call.** TTL 0 + `Cache-Control: no-store` eliminates all three failure modes (null-stickiness, cross-service mutation invalidation, zombie process) in one move.
- **CSRF design on the preview proxy is correct.** Allowing `same-origin` and `none`, rejecting `cross-site`/`same-site`, is precisely right for a `window.open` endpoint.
- **Slug validation blocks open-redirect.** `isValidTemplateGallerySlug` (`/^[a-z0-9]+(-[a-z0-9]+)*$/`) disallows `/`, `?`, `#`, `..`, uppercase, and non-ASCII — all attack shapes covered, with tests.
- **`proxy.ts` short-circuit is elegant** — the prefix guard ensures template-gallery paths route correctly regardless of hostname, with a test confirming no false prefix match.
- **`mountedRef` StrictMode fix is correct.** Adding `mountedRef.current = true` to the effect setup body is the right fix, and the best-practices doc explains the subtlety clearly.
- **`TEMPLATE_GALLERY_SLUG_RE` as a shared lib constant** prevents form validator and preview proxy from independently drifting on slug format rules.
- **Plan and solution docs are excellent.** The `docs/solutions/runtime-errors/` doc on cached-null stickiness and the zombie-process trap is institutional knowledge that will save hours on future investigations.

---

**Two items before merge:** fix the snackbar await ordering (🔴), and verify the spec file encoding (🟡). Everything else can be followed up separately.

**Comment by csiyang** [comment]
Thanks Mike — solid review. Shipped in c7e27d7bf (rebased onto the latest main merge as 7770fafdc).

## 🔴 Critical — applied

**Snackbar await ordering in TrashJourneyDialog + ArchiveJourney.** Flipped: snackbar now fires on the same tick as the primary mutation; `unassignFromCollection` is fire-and-forget with its own `.catch()` for the log line. Added a regression test in each dialog that holds the unassign mock at a never-resolving promise and asserts the snackbar still appears — closes the gap you flagged (waitFor doesn't assert ordering).

## 🟡 Warnings — applied

- **Binary spec file** — confirmed root cause was a literal NUL byte in the "null byte" test fixture, not CRLF. Rewrote `templateGallerySlug.spec.ts` using Unicode escape sequences; `file(1)` now reports plain ASCII and diffs render in PR tooling. 28 tests still passing.
- **Slug validation before Firebase auth** — reordered in `preview-template-gallery.ts`. New spec asserts `getApiRequestTokens` is never called for a malformed slug.
- **submittingRef silent drop** — added `showBusy()` helper that enqueues "Please wait for the current action to finish." when the guard fires, on all three mutation paths.
- **dragInFlight ref/state can drift** — wrapped in a single `setDragInFlight` setter in `TemplateGalleryPageList` that flips both atomically. `useDragEndHandler` no longer touches the ref directly; future callers can't desync them.
- **cache.modify duplication** — extracted `evictFromTemplateGalleryPages(cache, ids, { evictJourneyEntity })` into `apps/journeys-admin/src/libs/`. Both dialogs consume it (archive passes `evictJourneyEntity: false` so the Journey entity stays live in the archived list). New focused unit spec covers the helper directly.

## 🟢 Suggestions

- **publishBlockedReason JSDoc** — added explicit "callers MUST wrap in `t(...)` before rendering" note.
- **setTimeout(resolve, 200) in useCollectionMutations spec** — fair call but `jest.useFakeTimers()` would skip past the Apollo mock's real-timer delay too. Replaced with capturing the in-flight promise and awaiting it directly — deterministic regardless of CI load, no fake timers needed.
- **Redundant `slug == null` check in handleView** — skipped, this one was load-bearing. `viewDisabled` is just a boolean, so TypeScript can't narrow `slug` through it. Removing the runtime check made the type-system narrowing fail (`encodeURIComponent` doesn't accept `null`), and `slug as string` then tripped `@typescript-eslint/no-unnecessary-type-assertion`. Kept the check with a comment explaining it's the type-system pair, not a redundant safety net.

## Backend follow-up (out of scope here)

We discussed migrating `journeysTrash` and `journeysArchive` to api-journeys-modern so the join-row delete can happen transactionally inside the mutation — that would let the frontend drop the `…assignJourney({ pageId: null })` pairing entirely. Following your `chore: *modern` pattern. I'll file a separate ticket.

**Review by mikeallisonJS — DISMISSED** [review]
## Re-review: All original issues addressed ✅

All six items from the first review are cleanly resolved. A few minor notes below, none blocking.

---

### ✅ Issues from first review — all fixed

**🔴 Snackbar blocked by `await unassignFromCollection`** — Fixed correctly. Both `ArchiveJourney.tsx` and `TrashJourneyDialog.tsx` now use `void unassignFromCollection(...).catch(warn)` with the snackbar firing immediately. The regression test holds the unassign mock on a never-resolving promise to assert ordering directly — exactly the right test shape.

**🟡 `templateGallerySlug.spec.ts` binary file** — Fixed. Root cause was a literal NUL byte in a fixture string, not CRLF. Rewritten with Unicode escapes; 28 tests passing.

**🟡 Slug validation before Firebase auth** — Fixed, with a test explicitly asserting `mockGetApiRequestTokens` is never called for a malformed slug.

**🟡 `submittingRef` silent drop** — Fixed. `showBusy()` now fires an info snackbar when the entry guard blocks a concurrent mutation.

**🟡 `dragInFlight` ref/state divergence** — Fixed. A unified `setDragInFlight` setter flips both ref and state atomically; all call sites updated.

**🟡 `cache.modify` duplication** — Fixed. `evictFromTemplateGalleryPages` extracted to its own lib with JSDoc, direct unit tests, and the `evictJourneyEntity` option correctly differentiating trash vs. archive behavior.

---

### 🟡 Minor notes (non-blocking, file separately if needed)

**SSR + TTL 0 traffic confirmation** — The `yoga.ts` comment documents the reasoning well ("One indexed slug lookup per request is trivial"). Before this sees significant public traffic, it would be worth a quick note in Linear confirming the expected request volume for `/template-gallery/*` routes, so there's a documented baseline if server costs spike unexpectedly. Not a merge blocker given the current admin-managed, low-traffic nature of these pages.

**`mountedRef` comment clarity** — The StrictMode double-invoke explanation is correct but terse. A one-line note like "the setup body resets it after StrictMode cleanup leaves it `false`" would prevent a future reader from removing what looks like a redundant assignment.

**`useCanPublishCollection` error → fail-closed on transient network blip** — If the custom domains query gets a transient 500, all Publish/Preview CTAs immediately disable. Worth a follow-up issue to consider retry-with-backoff before switching to fail-closed, or at least a staleness window.

**`handlePreview` dead branch in `CollectionCard.tsx`** — The `if (previewDisabled) return` guard inside `handlePreview` is unreachable because MUI swallows clicks on disabled menu items. A short comment noting it's a safety net (not the primary guard) would prevent confusion.

---

**Approving.** The critical fix (snackbar ordering), the binary spec file, and all four warnings are fully resolved with correct tests. The non-blocking notes above are all candidates for follow-up issues rather than merge blockers.

**Review by mikeallisonJS — APPROVED** [review]
